CB-230: Migrate dump_manager.py off the ORM

ferbncode opened the pull request:
Migrated all functions i.e `public`, `import` and `json` of the `dump_manager.py` off the ORM.

ferbncode commented:
The dumps created from the `public` function are a bit different from the present [dumps](ftp://ftp.musicbrainz.org/pub/musicbrainz/critiquebrainz/dump) (Only the sequence of the columns in the `review` and `license`  have changed). Thus, the dumps [here](ftp://ftp.musicbrainz.org/pub/musicbrainz/critiquebrainz/dump) need to be updated as the `import` function is dependent on how the dumps are created.
Also I have updated the `model.__version__` to `db.SCHEMA_VERSION` and changed `db.SCHEMA_VERSION` to `6` (the present value of `model.__version__`). Is it preferred this way?

gentlecat commented on critiquebrainz/db/review.py:
```diff
@@ -575,3 +575,22 @@ def delete(review_id):
+    with db.engine.connect() as connection:
+        results = connection.execute(sqlalchemy.text("""
+            SELECT
+          DISTINCT entity_id
```
I might be a better to write it as `SELECT DISTINCT entity_id` on one line so that it's obvious what's being selected. It got me confused for a moment.

The goal is not to just split up by keywords, but to split up logical parts of a statement so that it's easier to read.

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -269,47 +321,68 @@ def importer(archive):
+def import_data(file_name, tablename, columns=None):
+
+    if tablename == "user":
+        tablename = '"user"'
```
Would it make sense to do that for all table names? That seems a bit too specific.

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -269,47 +321,68 @@ def importer(archive):
+                    raise Exception("Failed to copy data for table %s" % tablename)
+            else:
+                result = subprocess.call("""
+                    psql -h {} -p {} -U {} -d {} -a -c '\COPY {tablename} FROM {file_name}'
```
Does this work without specification of columns?

ferbncode replied:
Yes, it does, if data for all the columns of a table is present in the file and columns are in the same order as the table structure. The `public` function for making db dumps creates files ensuring the same order of columns.

ferbncode commented on critiquebrainz/db/__init__.py:
```diff
@@ -2,7 +2,7 @@
 
 # This value must be incremented after schema changes on exported tables!
-SCHEMA_VERSION = 1
+SCHEMA_VERSION = 6
```
I thought of updating the `db.SCHEMA_VERSION` to `6` as the previous value of `model.__version__` was `6`. I have changed it back to `1`.

gentlecat replied:
I see. I didn't quite understand where that number came from at first, which is why I asked. It should be the same as it was before actually (`6`). It needs to be like that so that there are no compatibility issues when someone tries to import data dump generated from an old schema.

ferbncode commented:
I have modified the PR and updated it to use raw database connections and cursors. It surely makes sense to use them instead of importing via `psql`😅.

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -93,26 +92,26 @@ def json(location, rotate=False):
             license_dir = os.path.join(temp_dir, safe_name)
             create_path(license_dir)
 
             # Finding release groups that have reviews with current license
```
Might be a good idea to update that comment since you are here. It's not just "release groups" anymore.

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -93,26 +92,26 @@ def json(location, rotate=False):
     print("Creating new archives...")
-    for license in model.License.query.all():
-        safe_name = slugify(license.id)
+    for license in db.license.list_licenses():
```
`db.license`, `db.review`, and others need to be explicitly imported.
```python
from critiquebrainz.db import license...
```

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -189,19 +189,25 @@ def public(location, rotate=False):
     # 1. COMBINED
     # Archiving all reviews (any license)
-    REVISION_COMBINED_SQL = "SELECT %s FROM revision JOIN review " \
+    REVISION_COMBINED_SQL = "SELECT revision.id, revision.review_id, timestamp, text " \
```
As far as I know consistent order of columns during export and import is important. That's what `get_columns` function handles. If the order is indeed important then that function should be replaced with something that would provide a similar functionality.

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -269,47 +282,59 @@ def importer(archive):
-def import_data(file_name, model, columns=None):
-    db_connection = db.session.connection().connection
-    cursor = db_connection.cursor()
+def import_data(file_name, tablename, columns=None):
```
`table_name`

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -269,47 +282,59 @@ def importer(archive):
-        import_data(os.path.join(temp_dir, 'cbdump', 'license'), model.License)
-        import_data(os.path.join(temp_dir, 'cbdump', 'review'), model.Review)
-        import_data(os.path.join(temp_dir, 'cbdump', 'revision'), model.Revision)
+        import_data(os.path.join(temp_dir, 'cbdump', 'user_sanitised'), '"user"',
```
I don't think `'"user"'` is the best way to make sure that keywords are properly quoted. When I use `import_data` function I probably shouldn't worry about that kind of thing. What if all column names are actually quoted like it was done before?
```python
cursor.copy_from(f, '"%s"' % model.__tablename__, columns=columns)
```

gentlecat commented on critiquebrainz/data/dump_manager.py:
```diff
@@ -269,47 +282,59 @@ def importer(archive):
+        count = cursor.fetchone()[0]
+        if count > 0:
+            print("Table %s already contains data. Skipping." % tablename)
+            return
```
This part can be moved into a separate function so that the checking is done like this:
```python
if has_data(table):
    print("Table %s already contains data. Skipping." % table)
    return
```

gentlecat commented:
OK, looks good now. Let me just try that with an updated base image.
